### PR TITLE
os/bluestore: optional SharedBlob on Blob structure

### PR DESCRIPTION
**Pull request by pereman2**
Blob::shared_blob was always set to something even though a SharedBlob wasn't really "created". In this PR we make shared_blob truly optional and only set it to a real instance if it's really necessary.

**Some results:**
```bash
bin/rados bench 10 write -p .mgr -t 4 --num-objects 20000 --no-cleanup
# with changes
"bluestore_blob_bytes": 4939272,
"bluestore_blob_items": 22867,
"bluestore_shared_blob_bytes": 0,
"bluestore_shared_blob_items": 0,

# without changes
"bluestore_blob_bytes": 4618200,
"bluestore_blob_items": 23091,
"bluestore_shared_blob_bytes": 554184,
"bluestore_shared_blob_items": 23091,
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by ifed01** [comment]
@pereman2 - please update the caption, it's confusing ;)

**Comment by pereman2** [comment]
```bash
bin/rados bench 10 write -p .mgr -t 4 --num-objects 20000 --no-cleanup
# with changes
"bluestore_blob_bytes": 4939272,
"bluestore_blob_items": 22867,
"bluestore_shared_blob_bytes": 0,
"bluestore_shared_blob_items": 0,

# without changes
"bluestore_blob_bytes": 4618200,
"bluestore_blob_items": 23091,
"bluestore_shared_blob_bytes": 554184,
"bluestore_shared_blob_items": 23091,
```

**Comment by petrutlucian94** [comment]
I see a potentially related OSD crash:

```
/home/ubuntu/ceph/src/os/bluestore/BlueStore.cc: 1428: FAILED ceph_assert(buffer_bytes >= b->length)

 ceph version cf68ec (ccf68ec392c31a754334e6538793e11c8b32572a) reef (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x15d) [0x560dcf628386]
 2: /home/ubuntu/ceph/build/bin/ceph-osd(+0x6275b5) [0x560dcf6285b5]
 3: (TwoQBufferCacheShard::_rm(BlueStore::Buffer*)+0x233) [0x560dcfdb7cf3]
 4: (BlueStore::BufferSpace::_rm_buffer(BlueStore::BufferCacheShard*, std::_Rb_tree_iterator<std::pair<unsigned int const, std::unique_ptr<BlueStore::Buffer, std::default_delete<BlueStore::Buffer> > > >)+0x3d) [0x560dcfd9986d]
 5: (BlueStore::BufferSpace::_discard(BlueStore::BufferCacheShard*, unsigned int, unsigned int)+0x9cd) [0x560dcfcdd5dd]
 6: (BlueStore::Blob::discard_unallocated(BlueStore::Collection*)+0xcf) [0x560dcfce572f]
 7: (BlueStore::_wctx_finish(BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection>&, boost::intrusive_ptr<BlueStore::Onode>&, BlueStore::WriteContext*, std::set<BlueStore::SharedBlob*, std::less<BlueStore::SharedBlob*>, std::allocator<BlueStore::SharedBlob*> >*)+0x151) [0x560dcfd4ae21]
 8: (BlueStore::_do_truncate(BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection>&, boost::intrusive_ptr<BlueStore::Onode>&, unsigned long, std::set<BlueStore::SharedBlob*, std::less<BlueStore::SharedBlob*>, std::allocator<BlueStore::SharedBlob*> >*)+0x31c) [0x560dcfd4ceac]
 9: (BlueStore::_truncate(BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection>&, boost::intrusive_ptr<BlueStore::Onode>&, unsigned long)+0xa7) [0x560dcfd4d297]
 10: (BlueStore::_txc_add_transaction(BlueStore::TransContext*, ceph::os::Transaction*)+0x1700) [0x560dcfd6e2c0]
 11: (BlueStore::queue_transactions(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, std::vector<ceph::os::Transaction, std::allocator<ceph::os::Transaction> >&, boost::intrusive_ptr<TrackedOp>, ThreadPool::TPHandle*)+0x2ef) [0x560dcfd75c9f]
 12: (non-virtual thunk to PrimaryLogPG::queue_transactions(std::vector<ceph::os::Transaction, std::allocator<ceph::os::Transaction> >&, boost::intrusive_ptr<OpRequest>)+0x53) [0x560dcf977663]
 13: (ReplicatedBackend::submit_transaction(hobject_t const&, object_stat_sum_t const&, eversion_t const&, std::unique_ptr<PGTransaction, std::default_delete<PGTransaction> >&&, eversion_t const&, eversion_t const&, std::vector<pg_log_entry_t, std::allocator<pg_log_entry_t> >&&, std::optional<pg_hit_set_history_t>&, Context*, unsigned long, osd_reqid_t, boost::intrusive_ptr<OpRequest>)+0x8a8) [0x560dcfbe0cf8]
 14: (PrimaryLogPG::issue_repop(PrimaryLogPG::RepGather*, PrimaryLogPG::OpContext*)+0x8ba) [0x560dcf8ea3ca]
 15: (PrimaryLogPG::execute_ctx(PrimaryLogPG::OpContext*)+0x11c5) [0x560dcf950235]
 16: (PrimaryLogPG::do_op(boost::intrusive_ptr<OpRequest>&)+0x382b) [0x560dcf95486b]
 17: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x1c9) [0x560dcf78bbd9]
 18: (ceph::osd::scheduler::PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x64) [0x560dcfac7b04]
 19: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x618) [0x560dcf7b35d8]
 20: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x40b) [0x560dcfeb145b]
 21: (ShardedThreadPool::WorkThreadSharded::entry()+0x14) [0x560dcfeb4614]
 22: (Thread::entry_wrapper()+0x54) [0x560dcfe9e2b4]
 23: /lib/x86_64-linux-gnu/libc.so.6(+0x94b43) [0x7fc4d9427b43]
 24: /lib/x86_64-linux-gnu/libc.so.6(+0x126a00) [0x7fc4d94b9a00]
```

https://jenkins.ceph.com/view/Windows/job/ceph-windows-pull-requests/29325/artifact/artifacts/cluster/ceph_logs/osd.2.log

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -713,6 +714,19 @@ class BlueStore : public ObjectStore,
       if (--nref == 0)
 	delete this;
     }
+    bool is_loaded() const {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h:717]: Better rename to e.g. is_shared_loaded() or something

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -5039,14 +5039,16 @@ void BlueStore::Collection::load_shared_blob(SharedBlobRef sb)
...
+  ceph_assert(!b->is_loaded());
+  ceph_assert(!b->shared_blob);
 
   // update blob
   bluestore_blob_t& blob = b->dirty_blob();
   blob.set_flag(bluestore_blob_t::FLAG_SHARED);
   // drop any unused parts, unlikely we could use them in future
   blob.clear_flag(bluestore_blob_t::FLAG_HAS_UNUSED);
   // update shared blob
+  b->shared_blob = new SharedBlob(sbid, this);
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc]: I don't think assertions should be removed as it ensures function is not used incorrectly

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -5181,19 +5183,23 @@ void BlueStore::Collection::split_cache(
 	cache->rm_blob();
 	dest->cache->add_blob();
 	SharedBlob* sb = b->shared_blob.get();
-	if (sb->coll == dest) {
+        b->collection = dest;
+	if (sb && sb->collection == dest) {
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc]: agree

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: My rationales are:
1. "ceph_assert(!b->is_shared_loaded)" is redundant if followed by "ceph_assert(!b->shared_blob);"
2. set_shared_blob() makes "ceph_assert(!b->shared_blob);" redundant too, see

void set_shared_blob(SharedBlobRef sb) {
      ceph_assert((bool)sb);
      ceph_assert(!shared_blob);
      ceph_assert(sb->collection = collection);
      shared_blob = sb;
      ceph_assert(get_cache());
    }
3.  Not having duplicate assertions on blob's content in Collection::make_blob_shared() results in a bit clearer/readable code with a better data encapsulation - frankly speaking why do Collection cares about blob's internals? Well I understand that's still the case in other places but let's make this one a bit better once we're on it...

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -5181,19 +5179,21 @@ void BlueStore::Collection::split_cache(
...
+        b->collection = dest;
+        if (sb) {
+          if (sb->collection == dest) {
+            ldout(store->cct, 20) << __func__ << "  already moved " << *sb
+              << dendl;
+            return;
+          }
+          ldout(store->cct, 20) << __func__ << "  moving " << *b << dendl;
+          ldout(store->cct, 20) << __func__ << "  moving " << *sb << dendl;
+          ldout(store->cct, 20) << __func__
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:5191]: Can be combined to the line above

**Comment by pereman2** [comment]
jenkins test api

**Comment by pereman2** [comment]
jenkins test make check

**Comment by ifed01** [comment]
jenkins test api

**Review by ifed01 — APPROVED** [review]

**Comment by aclamk** [comment]
jenkins test make check

**Comment by pereman2** [comment]
jenkins test make check

**Comment by ifed01** [comment]
@pereman2 - this is not passing make check. Please fix.

**Comment by yuriw** [comment]
@pereman2 I suspect this PR broke the build:

https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/75654//consoleFull

build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-625-g42d98088/rpm/el8/BUILD/ceph-19.0.0-625-g42d98088/src/os/bluestore/BlueStore.cc: In member function ‘void BlueStore::Blob::finish_write(uint64_t)’:
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/19.0.0-625-g42d98088/rpm/el8/BUILD/ceph-19.0.0-625-g42d98088/src/os/bluestore/BlueStore.cc:2935:22: error: ‘coll’ was not declared in this scope; did you mean ‘cosl’?
 2935 | #define dout_context coll->store->cct
      |                      ^~~~

**Comment by ljflores** [comment]
jenkins test api

**Comment by kamoltat** [comment]
@pereman2 Hi


```

**Comment by kamoltat** [comment]
@pereman2 can you let us know if the failed job I pasted in the comment above is related, so I can drop and retest Thanks!

**Comment by pereman2** [comment]
@kamoltat We are aware of this issue as it is related, we encountered it on a different teuthology job. We are working on it right now.

**Comment by kamoltat** [comment]
> @kamoltat We are aware of this issue as it is related, we encountered it on a different teuthology job. We are working on it right now.

Thank you for confirming!

**Comment by pereman2** [comment]
jenkins test api

**Review by aclamk — APPROVED** [review]

**Comment by amathuria** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomc58uc37eU1955-wip-yuri4-testing-2024-02-13-1546

**Pull request merged**